File: leetcode/easy/count-square-sum-triples.ts
// https://leetcode.com/problems/count-square-sum-triples/

// MINE: binary search
function countTriples(n: number): number {
  let ans = 0

  for (let i = n; i > 2; --i) {
    let left = 1,
      right = i - 1

    while (left < right) {
      if (left * left + right * right < i * i) {
        ++left
        continue
      }
      if (left * left + right * right > i * i) {
        --right
        continue
      }
      if (left * left + right * right === i * i) {
        ++ans
        ++left
        --right
      }
    }
  }

  return ans * 2
}

// ANSWER

File: leetcode/easy/count-square-sum-triples.py
class Solution:
    # MINE
    def countTriples(self, n: int) -> int:
        ans = 0

        for i in range(n, 2, -1):
            left = 1
            right = i-1

            while(left < right):
                if left*left + right*right < i*i:
                    left += 1
                    continue
                if left*left + right*right > i*i:
                    right -= 1
                    continue
                if left*left + right*right == i*i:
                    ans += 1
                    left += 1
                    right -= 1

        return ans*2

# >>> list(range(0, -10, -1))
# [0, -1, -2, -3, -4, -5, -6, -7, -8, -9]
